### ISSUE TITLE
opam2 switch create in parallel

### ISSUE THREAD

**Issue by avsm** [ocaml/opam#3200]
I find it increasingly useful to be able to do `opam switch create <compiler1> <compiler...>` in parallel for several usecases such as:

- setting up a jbuilder workspace on a multicore laptop
- container builds of ocaml/opam2 that have to build several compilers into one switch

I have a few wrapper shell scripts, but if the ability to generate multiple switches in one cli call were in opam, it would both be more convenient but also respect the `OPAMJOBS` value so that it doesn't kill the machines its scheduled on.

Right now the worst of the opam container builds takes >4 hours to complete due to the number of switches built in serial.

**Comment by AltGr** [comment]
With the current architecture, that would be quite tricky to implement: we only have parallelism at the sub-command level, and handling of command graphs at the switch level. This means we would need to add a whole new layer of multiplexing actions of multiple switches in the same graph¹.

So for the time being, a shell script would really the viable option. It should be the case, but you can confirm that 2.0.0 locking is fine-grained enough to allow it ? (1.2.2 certainly wasn't, so I'll be glad that my work on thoroughly improving locking is useful!) Then, I guess, since it's OCaml compiling, that you could use `-j 1` in this case to avoid too much forking. 

¹ There has been some experiments a while ago to operate on multiple switches as early as the solver, though, which would allow stuff like upgrading all switches in parallel, or installing a package on all switches, with a single command. Maybe it would be worth pursuing them ?)

**Comment by Drup** [comment]
> It should be the case, but you can confirm that 2.0.0 locking is fine-grained enough to allow it ?

I haven't tested recently, but it definitely was working at some point during pre-alpha.

> ¹ There has been some experiments a while ago to operate on multiple switches as early as the solver, though, which would allow stuff like upgrading all switches in parallel, or installing a package on all switches, with a single command. Maybe it would be worth pursuing them ?

I do not think it's urgent, but that would be nice.

**Comment by avsm** [comment]
> There has been some experiments a while ago to operate on multiple switches as early as the solver, though, which would allow stuff like upgrading all switches in parallel, or installing a package on all switches, with a single command. Maybe it would be worth pursuing them ?

Yes that sounds about the right timing. Shell scripts work fine until then.  Installing a package on all (or a selection of) switches would be perfect for the Jbuilder workspaces feature as well.

**Comment by AltGr** [comment]
Ok, putting that back on my list. Let's hope the solver will cope though :)

**Comment by avsm** [comment]
I put the quick'n'dirty shell script plugin up at https://github.com/avsm/opam-switches

**Comment by k4rtik** [comment]
@avsm how can your plugin be used?

Also, just bumping up this issue.
